### ISSUE TITLE
Method to run already notified handlers on failure

### ISSUE THREAD

**Issue by jsternberg** [ansible/ansible#10498]
We have a problem with our provisioning with Ansible at the current moment related to handlers. We use handlers to restart services after modifying configuration, but we can end up in a bad situation sometimes when a failure happens after the configuration has changed, but before the notifications have run.

```
- name: change configuration
  copy:
    src: important_file.conf
    dest: /path/to/important/location.conf
  notify: restart important service

- name: failure
  fail: msg="I sometimes fail, but sometimes not."
```

In this case, the important service never gets restarted. If you run the playbook again, the configuration won't be changed and it won't trigger the restart, but the playbook run will "succeed".

I'd like a way to configure Ansible so it will run any pending notifies before it ends with a failure message.

It looks like there was some original talk about this feature in #4443, #4766, and #4777. The last was a different request then this (always run handlers). We don't want to always run them, but only when it's appropriate.

**Comment by bcoca** [comment]
running the playbook with --force-handlers will run them even after task failures, currently there is no way to run them after a playbook has failed and was not run with this option.

**Comment by jsternberg** [comment]
Ah, I appear to have misunderstood the conversation in those issues. I thought it just forced all handlers to run whether or not the task changed or not. It appears there's no way to turn this on by default. One of the people was very adamant he didn't want it to be configurable in ansible.cfg for some reason, but it doesn't appear like always using this flag would harm anything.

Can we get an option for this in ansible.cfg that defaults to off?

**Comment by bcoca** [comment]
That is why i kept this ticket open as a feature idea, it has been on my wish list for a while.

**Comment by jder** [comment]
I'm running into this problem, too, but `--force-handlers` doesn't seem to solve this, because it still won't run the handlers on the nodes which failed. (It does run them on other nodes, but this still leaves the nodes which had failures in a bad state.) Is this the expected behavior?

**Comment by bcoca** [comment]
@jder it should run for nodes that fail a task and don't get to run in the rest of the playbook, but it won't work for nodes that are removed from play due to connection failures.

**Comment by jder** [comment]
Thanks, @bcoca. That doesn't seem to be what's happening though. With this playbook:

```
---
-
  hosts: all

  tasks:
    - command: /bin/true
      notify: do ping

    - command: /bin/false
      when: "'a' not in group_names"

  handlers:
    - name: do ping
      ping: 
```

With `--force-handlers`, only the hosts in group 'a' run the ping task (because they skip the failing task). 

I just updated to 1.9.0.1 to test this again, and I see the same behavior, but in addition it acts the same way even without `--force-handlers` (i.e. with or without the flag, the "do ping" task runs on hosts in group a). Maybe a bug? Do you want a separate issue about this?

**Comment by jder** [comment]
I've made a new bug for this: https://github.com/ansible/ansible/issues/10602

**Comment by jimi-c** [comment]
Just following up on old issues. Since the above was merged, is there any reason we're keeping this open?

**Comment by acasademont** [comment]
Yep, I think this should be closed, a bit misleading if #10618 is already merged

**Comment by jimi-c** [comment]
Thanks for following up, closing now.
